[AIRFLOW-3406] Implement an Azure CosmosDB operator

tmiller-msft opened the pull request:
Add an operator and hook to manipulate and use Azure
CosmosDB documents, including creation, deletion, and
updating documents and collections.

Includes sensor to detect documents being added to a
collection.

Make sure you have checked _all_ steps below.

### Jira

- [X] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3406
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:
An Azure CosmosDB  hook, operator, and sensor for manipulating documents and collections

### Tests

- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [X] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [X] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality

- [X] Passes `flake8`

XD-DENG commented on airflow/contrib/operators/azure_cosmos_insertdocument_operator.py:
```diff
@@ -0,0 +1,67 @@
+        hook = AzureCosmosDBHook(azure_cosmos_conn_id=self.azure_cosmos_conn_id)
+
+        # Create the DB if it doesn't already exist
+        hook.create_database(self.database_name)
```
Seems you're using a `try-except` to handle the case in which the DB already exists. Is there any reason that it can not be handled in a more explicit way?

Possibly better to be `Check if the DB already exists -> if yes, skip creatiing DB; if no, create`?

tmiller-msft replied:
Thanks! Updated to query for existing database and only creating when not existing rather than using exception flow.

XD-DENG commented on airflow/contrib/operators/azure_cosmos_insertdocument_operator.py:
```diff
@@ -0,0 +1,67 @@
+        hook.create_database(self.database_name)
+
+        # Create the collection as well
+        hook.create_collection(self.collection_name, self.database_name)
```
The same question as above. You're using a `try-except` to handle the case where the collection already exists.

Personally I think it may be better to handle it in a more explicit fashion.

tmiller-msft replied:
Thanks! Updated to query for existing database and only creating when not existing rather than using exception flow.

XD-DENG commented on airflow/contrib/sensors/azure_cosmos_sensor.py:
```diff
@@ -0,0 +1,68 @@
+        :param document_id: The ID of the target document.
+        :type query: str
+        :param azure_cosmos_conn_id: The connection ID to use
+                              when connecting to CosmosDB.
```
This line is not indented correctly. It will fail Sphinx rendering.

tmiller-msft replied:
Thanks, updated to mirror this parameter on the other classes, which also addresses this since it no longer wraps at all.

XD-DENG reviewed (commented):
A few comments for your information.

tmiller-msft committed:
[AIRFLOW-3406] Implement an Azure CosmosDB operator

Add an operator and hook to manipulate and use Azure
CosmosDB documents, including creation, deletion, and
updating documents and collections.

Includes sensor to detect documents being added to a
collection.

feng-tao commented on airflow/contrib/hooks/azure_cosmos_hook.py:
```diff
@@ -0,0 +1,287 @@
+
+        return self.cosmos_client
+
+    def get_database_name(self, database_name=None):
```
are this private function or public function? (if latter, good to have docstring)

tmiller-msft replied:
private, updated to reflect

feng-tao commented on airflow/contrib/hooks/azure_cosmos_hook.py:
```diff
@@ -0,0 +1,287 @@
+
+        return db_name
+
+    def get_collection_name(self, collection_name=None):
```
same

tmiller-msft replied:
private, updated to reflect

feng-tao commented on airflow/contrib/hooks/azure_cosmos_hook.py:
```diff
@@ -0,0 +1,287 @@
+        # We need to check to see if this database already exists so we don't try
+        # to create it twice
+        existing_database = list(self.get_conn().QueryDatabases({
+            "query": "SELECT * FROM r WHERE r.id=@id",
```
not familiar with cosmos, but is `r` a valid table which contains all the existing databases??

tmiller-msft replied:
Yes, this is the syntax for queries for cosmos

feng-tao commented on airflow/contrib/hooks/azure_cosmos_hook.py:
```diff
@@ -0,0 +1,287 @@
+            return None
+
+
+def GetDatabaseLink(database_id):
```
what are these function? snake case for function name

tmiller-msft replied:
Helpers, updated name casing

feng-tao commented on airflow/contrib/hooks/azure_cosmos_hook.py:
```diff
@@ -0,0 +1,287 @@
+    return "dbs/" + database_id
+
+
+def GetCollectionLink(database_id, collection_id):
```
same

tmiller-msft replied:
Helpers, updated name casing

feng-tao commented on airflow/contrib/hooks/azure_cosmos_hook.py:
```diff
@@ -0,0 +1,287 @@
+    return GetDatabaseLink(database_id) + "/colls/" + collection_id
+
+
+def GetDocumentLink(database_id, collection_id, document_id):
```
same

tmiller-msft replied:
Helpers, updated name casing

feng-tao reviewed (commented):
couples of small nits, but I don't know how it works as I don't have the env to test these operators/hooks etc.
